actionmailbox/app/controllers/action_mailbox/ingresses/amazon/inbound_emails_controller.rb

```diff
@@ -0,0 +1,137 @@
+  #          arn:aws:sns:eu-west-1:123456789001:example-topic-1
+  #          arn:aws:sns:us-east-1:123456789002:example-topic-2
+  #        )
+  #
+  # 4. {Configure SES}[https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-notifications.html]
+  #    to route emails through SNS.
+  #
+  #    Configure SNS to send emails to +/rails/action_mailbox/amazon/inbound_emails+.
+  #
+  #    If your application is found at <tt>https://example.com</tt> you would
+  #    specify the fully-qualified URL <tt>https://example.com/rails/action_mailbox/amazon/inbound_emails</tt>.
+  #
+  module Ingresses
+    module Amazon
+      class InboundEmailsController < ActionMailbox::BaseController
+        before_action :verify_authenticity
+        before_action :validate_topic
+        before_action :confirm_subscription
+
+        def create
+          head :bad_request unless mail.present?
+
+          ActionMailbox::InboundEmail.create_and_extract_message_id!(mail)
+          head :no_content
+        end
+
+        private
+          def verify_authenticity
+            head :bad_request unless notification.present?
+            head :unauthorized unless verified?
+          end
+
+          def confirm_subscription
+            return unless notification["Type"] == "SubscriptionConfirmation"
+            return head :ok if confirmation_response_code&.start_with?("2")
+
+            Rails.logger.error("SNS subscription confirmation request rejected.")
+            head :unprocessable_entity
+          end
+
+          def validate_topic
+            return if valid_topics&.include?(topic)
+
+            Rails.logger.warn("Ignoring unknown topic: #{topic}")
+            head :unauthorized
+          end
+
+          def confirmation_response_code
+            @confirmation_response_code ||= begin
+              Net::HTTP.get_response(URI(notification["SubscribeURL"])).code
+            end
+          end
+
+          def notification
+            @notification ||= JSON.parse(request.body.read)
+          rescue JSON::ParserError => e
+            Rails.logger.warn("Unable to parse SNS notification: #{e}")
+            nil
+          end
+
+          def verified?
+            verifier.authentic?(@notification.to_json)
+          end
+
+          def verifier
+            require "aws-sdk-sns"
+            Aws::SNS::MessageVerifier.new
+          end
+
+          def message
+            @message ||= JSON.parse(notification["Message"])
+          end
+
+          def mail
+            return nil unless notification["Type"] == "Notification"
+            return nil unless message["notificationType"] == "Received"
+
+            message["content"]
+          end
+
+          def topic
+            return nil unless notification.present?
+
+            notification["TopicArn"]
+          end
+
+          def valid_topics
+            ActionMailbox.amazon.subscribed_topics
+          end
+      end
+    end
+  end
+end
```

Comment:
Noticed that we’re burdening a single action with handling two different kinds of requests. Suggest splitting out `ActionMailbox::Ingresses::Amazon::ConfirmationsController#create`. We can use [constraints](https://edgeguides.rubyonrails.org/routing.html#advanced-constraints) to route requests with the same HTTP action/path to the different actions based on `request.params[:Type]`.

Reply:
Thanks - I did not know about this feature. Implemented here: https://github.com/bobf/rails/commit/ab3448f6b63f38b80987fa265c2b3024a98a3530